src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -1418,17 +1330,56 @@ private void UpdateBase2ExponentialHistogramWithMinMax(double number, ReadOnlySp
             histogram.RunningMax = Math.Max(histogram.RunningMax, number);
         }
 
-        if (reportExemplar && isSampled)
+        this.OfferExemplarIfSampled(number, tags, isSampled);
+
+        this.mpComponents.ReleaseLock();
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private readonly void OfferExemplarIfSampled<T>(T number, ReadOnlySpan<KeyValuePair<string, object?>> tags, bool isSampled)
+        where T : struct
+    {
+        if (isSampled)
         {
-            Debug.Assert(this.mpComponents.ExemplarReservoir != null, "ExemplarReservoir was null");
+            Debug.Assert(this.mpComponents?.ExemplarReservoir != null, "ExemplarReservoir was null");
 
             // TODO: Need to ensure that the lock is always released.
             // A custom implementation of `ExemplarReservoir.Offer` might throw an exception.
-            this.mpComponents.ExemplarReservoir!.Offer(
-                new ExemplarMeasurement<double>(number, tags));
+            if (typeof(T) == typeof(long))
+            {
+                this.mpComponents!.ExemplarReservoir!.Offer(
+                    new ExemplarMeasurement<long>((long)(object)number, tags));
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                this.mpComponents!.ExemplarReservoir!.Offer(
+                    new ExemplarMeasurement<double>((double)(object)number, tags));
+            }
```

Comment:
would this work?

```suggestion
            if (typeof(T) == typeof(long) || typeof(T) == typeof(double))
            {
                this.mpComponents!.ExemplarReservoir!.Offer(
                    new ExemplarMeasurement<T>(number, tags));
            }
```

Reply:
Sadly, no. `ExemplarReservoir.Offer` has two overloads:

```csharp
    public abstract void Offer(in ExemplarMeasurement<long> measurement);
    public abstract void Offer(in ExemplarMeasurement<double> measurement);
```

The code as you have it there the compiler doesn't know which one you want to call so it (more or less) guesses the first one and spits out:

> Error CS1503 Argument 1: cannot convert from 'OpenTelemetry.Metrics.ExemplarMeasurement<T>' to 'in OpenTelemetry.Metrics.ExemplarMeasurement<long>'